### PR TITLE
[WIP] Serve BIP 157 compact filters

### PR DESCRIPTION
**Pull request by jnewbery**
This is almost the same change set as #16442 but:

- fixup commits have been squashed into their appropriate commits
- some commits have been broken up into smaller pieces
- ordering of commits has changed
- a few other minor changes

This PR is marked WIP as it's a demonstration of the entire change. I'll split off commits into smaller PRs for merge.

The original PR has already received a lot of review, so I think it should be possible to get this merged over the next few weeks. To hold myself accountable, I'm targeting the following timeline. Obviously, the PRs shouldn't be merged until they've received thorough review.

| | Functionality | Commits | PR | Target merge date | Actual merge date |
| --- | --- | --- | --- | --- | --- |
| ✓ | Serve cfcheckpt requests | _[init] Add -peercfilters option_, _[net processing] Message handling for getcfcheckpt._, _[test] Add test for cfcheckpt_ | #18877 | 2020-05-13 | 2020-05-12 |
| ✓ | Cache cfcheckpt headers | _[net processing] Cache compact filter checkpoints in memory._ | #18960 | 2020-05-20 | 2020-05-21 |
| ✓ | Serve cfheaders requests | _[net processing] Message handling for getcfheaders._, _[test] Add test for cfheaders_ | #19010 | 2020-05-27 | 2020-05-26 |
| ✓ | Serve cfilters requests | _[indexes] Fix default [de]serialization of BlockFilter._, _[net processing] Message handling for getcfilters._, _[test] Add test for cfilters._ | #19044 | 2020-06-03 | 2020-05-31 |
| ✓ | Signal NODE_COMPACT_FILTERS support | _[net] Signal NODE_COMPACT_FILTERS if we're serving compact filters._, _[test] Add test for NODE_COMPACT_FILTER._ | #19070 | 2020-06-10 | 2020-08-13 |

### CODE REVIEW

**Comment by luke-jr** [comment]
>I think it should be possible to get this merged over the next few weeks. 

This has Concept NACKs and should not be merged.

**Comment by jnewbery** [comment]
@Empact @jamesob @talkless @sipa @fjahr @pinheadmz @jonasnick @jkczyz @practicalswift @marcinja @MarcoFalke @kilrau @TheBlueMatt - thank you for your reviews on the original PR. I'm going to try to respond to feedback quickly to make progress on this. Please help by reviewing the linked PRs!

**Comment by Empact** [comment]
Concept ACK

Current Travis linter error:
```
test/functional/p2p_cfilters.py:11:1: F401 'test_framework.messages.NODE_COMPACT_FILTERS' imported but unused
^---- failure generated from test/lint/lint-python.sh
```

**Comment by ariard** [comment]
Withdrawing my Concept NACK on the previous PR, ecosystem-wise, BIP157 is better that current light client solutions. Even if there is still long-term serious concerns, it's a broader point that this specific protocol. I can also envision people always servicing filters-headers only to provide filters correctness cheap services.

Few points, maybe for follow-up works:
*  signal NODE_COMPACT_FILTERS_HEADERS only, both filter-headers and filters are quite different in size, first one is the same order of magnitude that header and even a more resource-preserving peer still want to offer them
* it could be the opportunity to experiment on filters services some bandwidth DoS-protection, you may need to tie inbound peer addresses with some accounting logic to do so
* dissociating messages path from `net_processing` in the future, `ProcessMessages` may fan-out between classes of messages like `ProcessAddr`, `ProcessBlock`, `ProcessTransaction`, `ProcessFilters`. It would make easier to reason on their buffer and timers.

**Comment by michaelfolkson** [comment]
Concept ACK.

I need a stronger, more convincing argument that improved light client protocols will drive the number of independent parties running and using a listening full node down.

If I can be convinced of that I'd consider reversing the Concept ACK. It is clear that that metric is critically important to network health, I don't think there is any disagreement there. But I think a future where every mobile is a light client is just as likely to drive that number up.

I will continue to follow the discussion on the [mailing list](https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2020-May/017818.html).

**Comment by ariard** [comment]
@michaelfolkson for discussion clarity, and maybe it isn't clear in my mail sorry, I don't think in any case it will drive the number of independent parties running and using a listening full node down. If bandwidth cost is too high, node operators are just going to stop turn `peerscfilters=false`. The concern is for the light clients ecosystem in itself.

**Comment by michaelfolkson** [comment]
Sure, that was clear from your mail @ariard. I'm referring to some of the responses to your mailing list post which make arguments for this specific PR to be NACKed. All this project (Bitcoin Core) can do is facilitate superior light client protocols (assuming they are not harmful to the full node ecosystem which I haven't been convinced they are). If the light client ecosystem doesn't take off or if node operators all choose to turn `peerscfilters=false" there isn't much Core can do or should do. That is down to the node operators. But I'm trying to keep the arguments specific to this particular PR here. Any brainstorming on how we can make the light client ecosystem successful should *definitely* not be done here and should be done on the mailing list. 

Personally, I think Concept ACK, Concept NACKs are fine to post here with a short explanation. But if we can keep as much of the discussion on the mailing list as possible I think that is preferable.

**Comment by luke-jr** [comment]
>I need a stronger, more convincing argument that improved light client protocols will drive the number of independent parties running and using a listening full node down.

Currently, there is a strong incentive to use your own node for privacy. Widely-deployed on nodes, BIP157 destroys this incentive.

**Comment by michaelfolkson** [comment]
Running a full node will always be the superior option for privacy (and trust minimization) versus using a light client (improved or otherwise). If your primary objective is privacy the incentive is still there to run a full node. BIP157 reduces the gap between full node privacy and light client privacy that is true (light client privacy is improving). But this only affects a cohort of people who are wavering on whether to run a full node or not and see the privacy improvements of light clients as a reason not to. I have no idea how to quantify this cohort of people but I would be shocked if it was significant. Hence "destroys" seems to me to be a superlative.

On the upside, if the light client ecosystem was to grow and prosper post BIP157 it would need an increased number of full nodes to serve them filters. If the number of full nodes didn't increase to serve this filter demand then the light client ecosystem will be stunted forcing the cohort discussed above to lean back towards running a full node.

If the metric is "number of independent parties running and using a listening full node" I think growth of one ecosystem (light clients) has a positive impact on the growth of another (full nodes). If the metric is the **ratio** of light clients to "number of independent parties running and using a listening full node" then I think you have a point. I was of the understanding that we are more interested in the former than the latter.

**Comment by MarcoFalke** [comment]
re-𝔸ℂ𝕂 0ab77d46a0 🐯

Only change since my previous review https://github.com/bitcoin/bitcoin/pull/16442#issuecomment-623415561 is:

* Small simplification refactor in init.cpp
* Removing unused includes and symbols
* Add log categories to log statements
* Change return type to void where possible
* Rework documentation and test a bit 

Though, I was wondering why you dropped this hunk in the test:

```
# Check that nodes have signalled NODE_COMPACT_FILTERS correctly.
assert node0.nServices & NODE_COMPACT_FILTERS != 0
assert node1.nServices & NODE_COMPACT_FILTERS == 0

# Check that the localservices is as expected.
assert int(self.nodes[0].getnetworkinfo()['localservices'], 16) & NODE_COMPACT_FILTERS != 0
assert int(self.nodes[1].getnetworkinfo()['localservices'], 16) & NODE_COMPACT_FILTERS == 0
```

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-𝔸ℂ𝕂 0ab77d46a0 🐯

Only change since my previous review https://github.com/bitcoin/bitcoin/pull/16442#issuecomment-623415561 is:

* Small simplification refactor in init.cpp
* Removing unused includes and symbols
* Add log categories to log statements
* Change return type to void where possible
* Rework documentation and test a bit 

Though, I was wondering why you dropped this hunk in the test:

# Check that nodes have signalled NODE_COMPACT_FILTERS correctly.
assert node0.nServices & NODE_COMPACT_FILTERS != 0
assert node1.nServices & NODE_COMPACT_FILTERS == 0

# Check that the localservices is as expected.
assert int(self.nodes[0].getnetworkinfo()['localservices'], 16) & NODE_COMPACT_FILTERS != 0
assert int(self.nodes[1].getnetworkinfo()['localservices'], 16) & NODE_COMPACT_FILTERS == 0

-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUj5BgwAzu7Z+RNM0h2GPe3S/9ft5ZmqVl/U7OKuuwXNzoeMb9gk7RiM2HUPgZY0
q+8TxkMRtENG5MvIBJyGa6WvVdFGTTJMK93rLGAYdnq73lRj/D132cM6sxmbfOqt
AkBYl655mk73mPMaSbuSZfk1c4C+mXlWnk3W4dDgfMATclsttSf9fgzJ2nAaPUST
luw/ReUAzUiur4kXyDLX9CN8/VdnpAW9kan3RE5K1nyZ2uAPHkKqPfGff3TRp6l4
+mmHT6+d/H2MlQ8kH+0pjFKw+CNDpbi2rd9bJiNSJ9hDgnRzzC6B2q71VvoxCY43
+dYv541QqiTptd5PMByNN17d6kHiq9kCRlcs0t0YLGMkXDfB9aWxZp5jKRnhMa4Q
76LN1I5DUK7R8/qvdBgEOc7yqr2dfELba8wUlGpAeZNrsFpW3/Y50JeHZihpBJdY
KBgaKR1MU6uUjId7EAOpBz1YmXyRf0+2iNtH/syIW3opctzpNHaQHO6JiHOEF3qr
XeyEdd5x
=rpCM
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `cebd0143a8bfb0f77090cd044ec018bea0df73270bf57843fb892d5d6d9ac941  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108cebd0143a8bfb0f77090cd044ec018bea0df73270bf57843fb892d5d6d9ac941f010d45fee6cb4b94dcf8560ba3d8245f3dd08fff010030befb32ce2ef7d4c9be3800180cab808f1045eba9b04f0082ea30cf31e2484470083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff01018e38eac9cfc5d4a0e0204b5640f6fd308f1204373e97b0f07cb51bab2727ded5819d9af14f142c18dffe592b52b314d59ea9d08f1045eba9b03f00855bf022cee10be7c0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff0107356a2acf8d9dac002ac1f0e0f2a88c108f1045eba9b04f0083738d6ee41cee2270083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df010be08207a3a16a5cbcf26e1d26f5a8bc208f1045eba9b03f0085c12f8dc9716544d0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>

**Comment by jnewbery** [comment]
> Though, I was wondering why you dropped this hunk in the test:

That was removed for the first PR (since `NODE_COMPACT_FILTERS` is not defined until the last PR in the sequence). Thank you for reminding me to add it back to the final PR.

**Comment by jnewbery** [comment]
The next PR in the series is #18960: _Add compact block filter headers cache_

**Comment by jnewbery** [comment]
> > Though, I was wondering why you dropped this hunk in the test:

> That was removed for the first PR (since NODE_COMPACT_FILTERS is not defined until the last PR in the sequence). Thank you for reminding me to add it back to the final PR.

This has now been added back.

**Comment by jonasschnelli** [comment]
Concept ACK

**Comment by chris-belcher** [comment]
As of April 2019 the size of all the filters is 4GB. I'm imagining a smartphone having to download 4GB from a raspberry pi node just to start up its bitcoin wallet. Would SPV wallets actually use this technology? You can hardly call it "lightweight". Not to mention the wallet also has to download the blocks including false positives.

At 4GB it's not far off from needing less bandwidth to download the UTXO set.

Yes, Wasabi Wallet already uses filters like these, but the only reason they made it work is their wallet _only_ uses bech32 addresses, which are very rare on the blockchain. And if everyone used wallets that used those addresses then the filters would again become unusable large.

Obviously the filters are very useful for making wallets rescans faster (a fast machine does it in 10 hours right now), but filters being used by "lightweight" wallets seems like a pipe dream to me. Happy to be convinced otherwise.

**Comment by jonasschnelli** [comment]
@chris-belcher: it is a reasonable point. But does a light client really need to download the whole filterset (the 4GB)? If the app is freshly installed and a wallet has been created, is there a need to download anything from the past? Or would downloading the few MBs per week of filter data (and maybe some blocks including the false positives) be enough?

Things would obviously look different if one imports a key/seed/address/wallet with unknown birthday.

**Comment by chris-belcher** [comment]
If that's the plan then such a wallet won't be able to reasonably restore old seed phrases after a few years.

Sync'ing a brand new wallet has never been much of a problem I think, as even without filters the wallet could download all the new blocks (without witnesses) which would cost about 144 MB/day or just 1.7 kb/sec.

**Comment by prusnak** [comment]
Why would you download 144 MB/day if you could download 1MB/day? Think embedded/IoT devices, etc.

**Comment by ariard** [comment]
@chris-belcher if you assume this kind of chain backend are going to be primarily used for LN wallets, you may just bookmark confirmation of every of your funding outpoint. Even this kind of wallet may not expose at all onchain addresses. 

You already have to backup channel state, so their storage must already be stateful, you may just increase the scope of it to avoid rescan at all.

**Comment by MarcoFalke** [comment]
Just as a general note, I think we should use pull requests for code review of the code changes, not for general protocol and design discussions. Otherwise, it will be hard to follow the review process in the noise. Also, GitHub will start hiding comments if there are too many, making it painful to even read the code review and previous review comments. This can lead to review comments being overlooked and be left unaddressed.

**Comment by jnewbery** [comment]
The next PR in this sequence is #19010: _net processing: Add support for getcfheaders_

**Comment by jnewbery** [comment]
#19010 is merged. The next PR in this sequence is #19044: _net processing: Add support for getcfilters_

**Comment by jnewbery** [comment]
#19044 is merged. Next PR in the sequence is #19070: _Signal NODE_COMPACT_FILTERS support_.

**Comment by kilrau** [comment]
Forgive me if this was discussed before, but do the served cfilters count against `maxuploadtarget`? https://github.com/bitcoin/bitcoin/pull/16442#issuecomment-623416425 sounds like it.

**Comment by jnewbery** [comment]
rockets

**Comment by jnewbery** [comment]
> do the served cfilters count against maxuploadtarget? #16442 (comment) sounds like it.

Yes, the number of bytes will be counted towards the upload target (but peers will still be able to download compact filters after the target is reached, just as they can still download transactions).
